src/extremeweatherbench/utils.py

```diff
@@ -403,3 +320,176 @@ def read_event_yaml(input_pth: str | Path) -> dict:
+        max_lat: maximum latitude.
+        min_lon: minimum longitude.
+        max_lon: maximum longitude.
+
+    Returns a subset dataframe."""
+    location_subset_df = df[
+        (df["latitude"] >= min_lat)
+        & (df["latitude"] <= max_lat)
+        & (df["longitude"] >= min_lon)
+        & (df["longitude"] <= max_lon)
+    ]
+    return location_subset_df
+
+
+def align_point_obs_from_gridded(
+    forecast_ds: xr.Dataset,
+    case_subset_point_obs_df: pd.DataFrame,
+    data_var: List[str],
+    point_obs_metadata_vars: List[str],
+) -> Tuple[xr.Dataset, xr.Dataset]:
+    """Takes in a forecast dataarray and point observation dataframe, aligning them by
+    reducing dimensions.
+
+    Args:
+        forecast_ds: The forecast dataset.
+        case_subset_point_obs_df: The point observation dataframe.
+        data_var: The variable to subset (e.g. "surface_air_temperature")
+        point_obs_metadata_vars: The metadata variables to subset (e.g. ["elev", "name"])
+
+    Returns a tuple of aligned forecast and observation dataarrays."""
+    case_subset_point_obs_df = case_subset_point_obs_df[
+        POINT_OBS_METADATA_VARS + data_var
+    ]
+    case_subset_point_obs_df = case_subset_point_obs_df.rename(columns=ISD_MAPPING)
+    case_subset_point_obs_df["longitude"] = convert_longitude_to_360(
+        case_subset_point_obs_df["longitude"]
+    )
+    case_subset_point_obs_df = location_subset_point_obs(
+        case_subset_point_obs_df,
+        forecast_ds["latitude"].min().values,
+        forecast_ds["latitude"].max().values,
+        forecast_ds["longitude"].min().values,
+        forecast_ds["longitude"].max().values,
+    )
+    # Uses indexing in the dataframe to capture metadata columns for future use
+    point_obs_metadata = case_subset_point_obs_df[point_obs_metadata_vars]
+
+    # Reset index to allow for easier modification
+    case_subset_point_obs_df = case_subset_point_obs_df.reset_index()
+    case_subset_point_obs_df.loc[:, "station"] = case_subset_point_obs_df[
+        "station"
+    ].astype("str")
+
+    # Set up multiindex to enable slicing along individual timesteps
+    case_subset_point_obs_df = case_subset_point_obs_df.set_index(
+        ["station", "time"]
+    ).sort_index()
+
+    aligned_forecast_list = []
+    aligned_observation_list = []
+
+    # Loop init and lead times to prevent indexing error from duplicate valid times
+    for init_time, lead_time in itertools.product(
+        forecast_ds.init_time, forecast_ds.lead_time
+    ):
+        valid_time = pd.Timestamp(
+            init_time.values + pd.to_timedelta(lead_time.values, unit="h").to_numpy()
+        )
+        valid_time_index = pd.IndexSlice[:, valid_time]
+        obs_timeslice = (
+            case_subset_point_obs_df.loc[valid_time_index, :]
+            if valid_time in case_subset_point_obs_df.index.get_level_values(1)
+            else None
+        )
+        if obs_timeslice is None:
+            continue
```

Comment:
Can't you just take `obs_timeslice = case_subset_point_obs_df.loc[valid_time_index, :]` and wrap it in an `IndexError` or expect to take the zero-length DataFrame as the null case, rather than manually handle it by setting to `None`? It's best to use the default or empty value of an output as the base case when possible.

Reply:
Would this suffice? I don't want to break the code in this case, just keep looping: 

```suggestion
try:
    if valid_time in case_subset_point_obs_df.index.get_level_values(1):
        obs_timeslice = case_subset_point_obs_df.loc[valid_time_index, :]
    else:
        continue
```